### PR TITLE
Introduce `DiscoverySelectors.selectMethod(Class<?>, String, Class<?>...)`

### PR DESCRIPTION
**Pull request by sbrannen**
## Overview

This is a _draft_ PR for #3326.

Still need to introduce a `selectNestedMethod()` overload that accepts var-arg parameter types.

---

### Definition of Done

- [ ] There are no TODOs left in the code
- [ ] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [ ] [Coding conventions](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [ ] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [ ] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [ ] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)

### CODE REVIEW

**Comment by sbrannen** [comment]
@marcphilipp, as discussed, here is the draft PR for #3326.

It combines several sets of changes which I included in a single commit thus far. 

For the sake of time, I opted not to introduce separate commits since doing so is a bit cumbersome. Though, if that's too cumbersome to review, let me know, and I can introduce separate commits for the following (which this PR currently contains).

- Restructure `DiscoverySelectorsTests`: use `@Nested` test classes for logical grouping.
- Refactor `MethodSelector` constructors and `DiscoverySelectors.selectMethod()` methods so that the `ClassLoader` argument is first.
- Introduce `getParameterTypeNames()` and deprecate `getMethodParameterTypes()` in `MethodSelector` and `NestedMethodSelector`.
- Introduce `DiscoverySelectors.selectMethod(Class<?>, String, Class<?>...)`, `DiscoverySelectors.selectMethod(String, String, Class<?>...)`, and `DiscoverySelectors.selectMethod(ClassLoader, String, String, Class<?>...)`.

**Comment by sbrannen** [comment]
Related discussion on the actual commit instead of this PR: https://github.com/sbrannen/junit5/commit/b31465d3ca099f453fbdddf2605cf1e06ca164cf#r115749361

/cc @kriegfrj

File: junit-platform-commons/src/main/java/org/junit/platform/commons/util/ReflectionUtils.java [hunk 1]
```diff
@@ -1361,7 +1361,7 @@ public static Optional<Method> findMethod(Class<?> clazz, String methodName, Str
 		return findMethod(clazz, methodName, resolveParameterTypes(clazz, methodName, parameterTypeNames));
 	}
 
-	private static Class<?>[] resolveParameterTypes(Class<?> clazz, String methodName, String parameterTypeNames) {
+	public static Class<?>[] resolveParameterTypes(Class<?> clazz, String methodName, String parameterTypeNames) {
```
Comment (marcphilipp) [inline junit-platform-commons/src/main/java/org/junit/platform/commons/util/ReflectionUtils.java:1364]: Shouldn't this get `@API(status = INTERNAL, since = "1.10")`?

File: junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/DiscoverySelectors.java [hunk 2]
```diff
@@ -592,6 +592,68 @@ public static MethodSelector selectMethod(Class<?> javaClass, String methodName,
...
+	 * is declared, or a subclass thereof; never {@code null} or blank
+	 * @param methodName the name of the method to select; never {@code null} or blank
+	 * @param parameterTypes the formal parameter types of the method; never
+	 * {@code null} though potentially empty if the method does not declare parameters
+	 * @since 1.10
+	 * @see MethodSelector
+	 */
+	@API(status = EXPERIMENTAL, since = "1.10")
+	public static MethodSelector selectMethod(ClassLoader classLoader, String className, String methodName,
+			Class<?>... parameterTypes) {
```
Comment (marcphilipp) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/DiscoverySelectors.java:628]: I'm unsure about the overloads that take a `ClassLoader`, a `String` class name and a `Class<?>...` argument (such as here and below). Since the class is expected to be loaded by the supplied class loader it seems inconsistent. Did you add these overloads to supported mixed mode when indirectly called via the `@SelectMethod` annotation (in a future PR)?

**Review by marcphilipp — APPROVED** [review]

File: junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/DiscoverySelectors.java (hunk 2 above)
Comment (sbrannen) [reply junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/DiscoverySelectors.java:628]: > Since the class is expected to be loaded by the supplied class loader it seems inconsistent. 

Yes, the `Class` will be loaded by the supplied `ClassLoader`.

But I don't understand what you think is inconsistent.

Can you please expound?

> Did you add these overloads to supported mixed mode when indirectly called via the `@SelectMethod` annotation (in a future PR)?

I added these overloads to align with other changes we have made in 1.10 to support the general "explicit ClassLoader for selectors" theme.

If we don't have these overloads, the user has no control over how the `Class` is loaded for the supplied `className`.

In other words, this aligns with the impetus for introducing `DiscoverySelectors.selectMethod(ClassLoader, String, String, String)` which has also been introduced in 1.10.

Having said that, I do concede that these particular overloads may not be used by many people. I added them for the sake of completeness, since I figured it wouldn't hurt. Although it does increase the number of `selectMethod(...)` overloads. 

If you feel that we should not introduce these overloads, I can remove them, and we can add them later if requested by the community.

Comment (sbrannen) [reply junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/DiscoverySelectors.java:628]: Thanks for the clarification.

I'll remove those variants.

If we ever decide we need them, they can be recovered from the commit history in this PR.

Comment (sbrannen) [reply junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/DiscoverySelectors.java]: Done in 74e78639ccc83d9bcaac600fb5563c9d5054172e

**Comment by sbrannen** [comment]
@marcphilipp, I made the requested changes.

Do you want to take a final look before I merge this PR?

**Comment by marcphilipp** [comment]
@sbrannen Sorry, I was offline for much of the weekend and didn't see this until now.

**Comment by sbrannen** [comment]
No worries.... I just merged it. 🙂

File: junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/DiscoverySelectors.java (hunk 2 above)
Comment (kriegfrj) [reply junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/DiscoverySelectors.java]: > I meant that the test class will be loaded by the supplied class loader while the parameter types won't. If we don't have a use case for these right now I'd therefore omit them.

Sorry for being late to this discussion... this is a little bit complicated, but for posterity/future reference:

While it *can* be a problem if the parameters are loaded by a different classloader, it not *necessarily* a problem. In fact, in a multiple-classloader environment like OSGi, there are common circumstances where the parameter types *must* be loaded by a different classloader. Consider, for example, if you have a test class in a custom bundle - it will be loaded by the bundle's classloader. However, if it has a test method which takes a Java primitive as a parameter like `java.lang.String`, `java.lang.int` or `java.lang.Integer`, then the parameter will be loaded by the system classloader.

The more important question rather than "is it a different classloader" is "are the classes visible to the test class' classloader". The answer to this question in general depends the specific classloader hierarchy/environment, but it is certainly possible with the `(ClassLoader, String, String, Class<?>...)` variant to end up passing `Class` instances as parameters that are not visible to the test class' classloader and hence won't load. The same is also true of the `(ClassLoader, String, String, String...)` variant though, as you could pass in FQN of parameter types that are not visible to the `ClassLoader` you've provided.

Comment (sbrannen) [reply junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/DiscoverySelectors.java]: Thanks for chiming in, @kriegfrj.

> The answer to this question in general depends the specific classloader hierarchy/environment

I think that pretty much sums it up: _it depends_. 😉 

In any case, we'll go with what we have in `main` for the time being, and we can revisit this later if necessary.
